plugin fails with NullPointerException

1. Podman fails with NullPointerException when `imageName` tag is not specified for a `stage`. I thought that if `customImageNameForMultiStageContainerfile` is false, it is not required to specify the stage specific image name. Looks like a bug to me. If I specify `imageName`, then no exception occurs.

      ```xml
    <name>test-app/%a</name>
    <customImageNameForMultiStageContainerfile>false</customImageNameForMultiStageContainerfile>
    <stages>
        <stage>
            <name>pre-built</name>
        </stage>
    </stages>
      ```

1. The stage names printed in the debug log shows duplicate stage names. It turned out that in my Containerfile, I had a commented line like `# FROM quay.io/centos/centos:stream8 as healthcheck-builder`.
  
  Below is the log for both:
  ```bash
  Apache Maven 3.8.7 (b89d5959fcde851dcb1c8946a785a163f14e1e29)
  Maven home: /usr/local/etc/apache-maven-3.8.7
  Java version: 11.0.19, vendor: Red Hat, Inc., runtime: /usr/lib/jvm/java-11-openjdk-11.0.19.0.7-1.fc38.x86_64
  Default locale: en_US, platform encoding: UTF-8
  OS name: "linux", version: "6.3.12-200.fc38.x86_64", arch: "amd64", family: "unix"
  
  ...
  
  [INFO] --- podman-maven-plugin:1.17.0:build (podman-build) @ my-maven-module ---
  [DEBUG] Configuring mojo nl.lexemmens:podman-maven-plugin:1.17.0:build from plugin realm ClassRealm[plugin>nl.lexemmens:podman-maven-plugin:1.17.0, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@55054057]
  [DEBUG] Configuring mojo 'nl.lexemmens:podman-maven-plugin:1.17.0:build' with basic configurator -->
  [DEBUG]   (f) failOnMissingContainerfile = true
  [DEBUG]   (f) name = test/%a
  [DEBUG]   (s) customImageNameForMultiStageContainerfile = false
  [DEBUG]   (s) name = pre-built
  [DEBUG]   (s) stages = [nl.lexemmens.podman.config.image.StageConfiguration@70136c67]
  [DEBUG]   (s) containerFileDir = /home/kumadee/git/dev/test_app
  [DEBUG]   (s) noCache = true
  [DEBUG]   (s) format = OCI
  [DEBUG]   (s) tagWithMavenProjectVersion = true
  [DEBUG]   (s) args = {ARTIFACT_ID=my-maven-module, ARTIFACT_LOCAL_PATH=./target/my-maven-module-54-SNAPSHOT.jar, ARTIFACT_VERSION=54-SNAPSHOT, BASEIMAGE_REGISTRY=quay.io, CID=CID5352}
  [DEBUG]   (s) build = nl.lexemmens.podman.config.image.single.SingleImageBuildConfiguration@5e7214b7
  [DEBUG]   (f) images = [nl.lexemmens.podman.config.image.single.SingleImageConfiguration@334674b5]
  [DEBUG]   (f) project = MavenProject: org.example.test:my-maven-module:54-SNAPSHOT @ /home/kumadee/git/dev/test_app/my-maven-module/pom.xml
  [DEBUG]   (f) pushRegistry = quay.io
  [DEBUG]   (f) registries = []
  [DEBUG]   (f) settings = org.apache.maven.execution.SettingsAdapter@20134094
  [DEBUG]   (f) skip = false
  [DEBUG]   (f) skipAuth = true
  [DEBUG]   (f) skipBuild = false
  [DEBUG]   (f) skipCatalog = false
  [DEBUG]   (f) skipTag = false
  [DEBUG] -- end configuration --
  [DEBUG] Initializing configurations.
  [DEBUG] Using default Podman configuration.
  [DEBUG] Setting TLS Verify to NOT_SPECIFIED
  [DEBUG] Using Podman's default settings for --root.
  [DEBUG] Using Podman's default settings for --runroot.
  [DEBUG] Executing Podman from the module's directory.
  [DEBUG] Using default skopeo configuration.
  [DEBUG] Found a stage named: healthcheck-builder
  [DEBUG] Found a stage named: healthcheck-builder
  [DEBUG] Found a stage named: base-environment
  [DEBUG] Found a stage named: builder
  [DEBUG] Found a stage named: src
  [DEBUG] Found a stage named: debug
  [DEBUG] Found a stage named: artifactory
  [DEBUG] Found a stage named: pre-built
  [INFO] ------------------------------------------------------------------------
  [INFO] BUILD FAILURE
  [INFO] ------------------------------------------------------------------------
  [INFO] Total time:  33.925 s
  [INFO] Finished at: 2023-08-18T13:38:13+02:00
  [INFO] ------------------------------------------------------------------------
  [ERROR] Failed to execute goal nl.lexemmens:podman-maven-plugin:1.17.0:build (podman-build) on project my-maven-module: Execution podman-build of goal nl.lexemmens:podman-maven-plugin:1.17.0:build failed.: NullPointerException -> [Help 1]
  org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal nl.lexemmens:podman-maven-plugin:1.17.0:build (podman-build) on project my-maven-module: Execution podman-build of goal nl.lexemmens:podman-maven-plugin:1.17.0:build failed.
      at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:375)
      at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
      at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
      at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:298)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
      at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
      at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
      at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
      at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
      at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
      at java.lang.reflect.Method.invoke (Method.java:566)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
      at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
      at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
  Caused by: org.apache.maven.plugin.PluginExecutionException: Execution podman-build of goal nl.lexemmens:podman-maven-plugin:1.17.0:build failed.
      at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
      at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
      at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
      at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
      at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
      at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
      at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:298)
      at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
      at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
      at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
      at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
      at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
      at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
      at java.lang.reflect.Method.invoke (Method.java:566)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
      at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
      at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
      at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
  Caused by: java.lang.NullPointerException
  [ERROR] 
  [ERROR] 
  [ERROR] For more information about the errors and possible solutions, please read the following articles:
  [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
  ```

--- a/src/main/java/nl/lexemmens/podman/config/image/AbstractImageBuildConfiguration.java
+++ b/src/main/java/nl/lexemmens/podman/config/image/AbstractImageBuildConfiguration.java
@@ -370,7 +370,8 @@ public abstract class AbstractImageBuildConfiguration {
 
     protected void determineBuildStages(Log log, Path fullContainerFilePath) throws MojoExecutionException {
         try (Stream<String> containerFileStream = Files.lines(fullContainerFilePath)) {
-            List<String> content = containerFileStream.collect(Collectors.toList());
+            List<String> content = containerFileStream.filter(c -> !c.matches("^\\s*#.*"))
+                                                      .collect(Collectors.toList());
             for (String line : content) {
                 Matcher matcher = MULTISTAGE_CONTAINERFILE_REGEX.matcher(line);
                 if (matcher.find()) {

--- a/src/main/java/nl/lexemmens/podman/helper/ImageNameHelper.java
+++ b/src/main/java/nl/lexemmens/podman/helper/ImageNameHelper.java
@@ -63,8 +63,13 @@ public class ImageNameHelper {
         String imageName = parameterReplacer.replace(imageConfiguration.getImageName());
         imageConfiguration.setImageName(imageName);
 
-        if (imageConfiguration.getStages() != null && imageConfiguration.getStages().length > 0) {
+        if (imageConfiguration.useCustomImageNameForMultiStageContainerfile() &&
+                imageConfiguration.getStages() != null && imageConfiguration.getStages().length > 0) {
             for (StageConfiguration stage : imageConfiguration.getStages()) {
+                if (stage.getImageName() == null) {
+                    stage.setImageName(imageName);
+                    continue;
+                }
                 String stageImageName = parameterReplacer.replace(stage.getImageName());
                 stage.setImageName(stageImageName);
             }

--- a/src/test/java/nl/lexemmens/podman/BuildMojoTest.java
+++ b/src/test/java/nl/lexemmens/podman/BuildMojoTest.java
@@ -628,6 +628,126 @@ public class BuildMojoTest extends AbstractMojoTest {
     }
 
     @Test
+    public void testMultiStageBuildWithCustomTagPerStageIsNotConfigured() throws MojoExecutionException, IOException, URISyntaxException {
+        URI sampleBuildOutputUri = PushMojoTest.class.getResource("/multistagecontainerfile/samplebuildoutput.txt").toURI();
+        Path sampleBuildOutputPath = Paths.get(sampleBuildOutputUri);
+
+        List<String> buildOutputUnderTest;
+        try (Stream<String> buildSampleOutput = Files.lines(sampleBuildOutputPath)) {
+            buildOutputUnderTest = buildSampleOutput.collect(Collectors.toList());
+        }
+
+        Assertions.assertNotNull(buildOutputUnderTest);
+
+        PodmanConfiguration podman = new TestPodmanConfigurationBuilder().setTlsVerify(TlsVerify.FALSE).build();
+        SingleImageConfiguration image = new TestSingleImageConfigurationBuilder("sample")
+                .setContainerfileDir("src/test/resources/multistagecontainerfile")
+                .setTags(new String[]{"0.2.1"})
+                .setCreateLatestTag(false)
+                .setUseCustomImageNameForMultiStageContainerfile(false)
+                .addCustomImageNameForBuildStage("phase", "image-name-number-1")
+                .build();
+        configureMojo(podman, image, true, false, false, false, true);
+
+        when(mavenProject.getBuild()).thenReturn(build);
+        when(build.getDirectory()).thenReturn("target");
+        when(serviceHubFactory.createServiceHub(isA(Log.class), isA(MavenProject.class), isA(MavenFileFilter.class), isA(PodmanConfiguration.class), isA(SkopeoConfiguration.class), isA(Settings.class), isA(SettingsDecrypter.class), isA(MavenProjectHelper.class))).thenReturn(serviceHub);
+        when(serviceHub.getContainerfileDecorator()).thenReturn(containerfileDecorator);
+        when(serviceHub.getPodmanExecutorService()).thenReturn(podmanExecutorService);
+        when(serviceHub.getSecurityContextService()).thenReturn(securityContextService);
+        when(serviceHub.getMavenProjectHelper()).thenReturn(mavenProjectHelper);
+        when(podmanExecutorService.build(isA(SingleImageConfiguration.class))).thenReturn(buildOutputUnderTest);
+
+        buildMojo.execute();
+
+        // Verify logging
+        verify(log, times(1)).info("Detected multistage Containerfile...");
+
+        // At random verify some lines
+        verify(log, times(1)).debug("Processing line: 'STEP 1: FROM nexus.example:15000/adoptopenjdk/openjdk11:11.0.3 AS base'");
+        verify(log, times(1)).debug("Processing candidate: 'STEP 7: LABEL Build-User=sample-user2 Git-Repository-Url=null'");
+
+        // Verify stage detection
+        verify(log, times(1)).debug("Processing stage in Containerfile: base");
+        verify(log, times(1)).debug("Processing stage in Containerfile: phase");
+        verify(log, times(1)).debug("Processing stage in Containerfile: phase2");
+
+        // Verify hashes for stages
+        verify(log, times(1)).info("Final image for stage base is: 7e72c870614");
+        verify(log, times(1)).info("Final image for stage phase is: 7f55eab001a");
+        verify(log, times(1)).info("Final image for stage phase2 is: d2efc6645cb");
+
+        // Verify warning markers
+        verify(log, times(0)).warn("No image name configured for build stage: phase. Image 7f55eab001a not tagged!");
+        verify(log, times(0)).warn("No image name configured for build stage: phase2. Image d2efc6645cb not tagged!");
+
+        verify(log, times(0)).info("Tagging container image 7f55eab001a from stage phase as registry.example.com/image-name-number-1:0.2.1");
+        verify(log, times(1)).info("Tagging container image d2efc6645cbb6ea012f8adcaaab6b03ef847dd3d2b4fa418ca4cde57eff28a7f as registry.example.com/sample:0.2.1");
+
+        verify(log, times(1)).info("Built container image.");
+    }
+
+    @Test
+    public void testMultiStageBuildWithCustomTagPerStageIsConfiguredButStageImageNameIsMissing() throws MojoExecutionException, IOException, URISyntaxException {
+        URI sampleBuildOutputUri = PushMojoTest.class.getResource("/multistagecontainerfile/samplebuildoutput.txt").toURI();
+        Path sampleBuildOutputPath = Paths.get(sampleBuildOutputUri);
+
+        List<String> buildOutputUnderTest;
+        try (Stream<String> buildSampleOutput = Files.lines(sampleBuildOutputPath)) {
+            buildOutputUnderTest = buildSampleOutput.collect(Collectors.toList());
+        }
+
+        Assertions.assertNotNull(buildOutputUnderTest);
+
+        PodmanConfiguration podman = new TestPodmanConfigurationBuilder().setTlsVerify(TlsVerify.FALSE).build();
+        SingleImageConfiguration image = new TestSingleImageConfigurationBuilder("sample")
+                .setContainerfileDir("src/test/resources/multistagecontainerfile")
+                .setTags(new String[]{"0.2.1"})
+                .setCreateLatestTag(false)
+                .setUseCustomImageNameForMultiStageContainerfile(true)
+                .addCustomImageNameForBuildStage("phase", null)
+                .build();
+        configureMojo(podman, image, true, false, false, false, true);
+
+        when(mavenProject.getBuild()).thenReturn(build);
+        when(build.getDirectory()).thenReturn("target");
+        when(serviceHubFactory.createServiceHub(isA(Log.class), isA(MavenProject.class), isA(MavenFileFilter.class), isA(PodmanConfiguration.class), isA(SkopeoConfiguration.class), isA(Settings.class), isA(SettingsDecrypter.class), isA(MavenProjectHelper.class))).thenReturn(serviceHub);
+        when(serviceHub.getContainerfileDecorator()).thenReturn(containerfileDecorator);
+        when(serviceHub.getPodmanExecutorService()).thenReturn(podmanExecutorService);
+        when(serviceHub.getSecurityContextService()).thenReturn(securityContextService);
+        when(serviceHub.getMavenProjectHelper()).thenReturn(mavenProjectHelper);
+        when(podmanExecutorService.build(isA(SingleImageConfiguration.class))).thenReturn(buildOutputUnderTest);
+
+        buildMojo.execute();
+
+        // Verify logging
+        verify(log, times(1)).info("Detected multistage Containerfile...");
+
+        // At random verify some lines
+        verify(log, times(1)).debug("Processing line: 'STEP 1: FROM nexus.example:15000/adoptopenjdk/openjdk11:11.0.3 AS base'");
+        verify(log, times(1)).debug("Processing candidate: 'STEP 7: LABEL Build-User=sample-user2 Git-Repository-Url=null'");
+
+        // Verify stage detection
+        verify(log, times(1)).debug("Processing stage in Containerfile: base");
+        verify(log, times(1)).debug("Processing stage in Containerfile: phase");
+        verify(log, times(1)).debug("Processing stage in Containerfile: phase2");
+
+        // Verify hashes for stages
+        verify(log, times(1)).info("Final image for stage base is: 7e72c870614");
+        verify(log, times(1)).info("Final image for stage phase is: 7f55eab001a");
+        verify(log, times(1)).info("Final image for stage phase2 is: d2efc6645cb");
+
+        // Verify warning markers
+        verify(log, times(0)).warn("No image name configured for build stage: phase. Image 7f55eab001a not tagged!");
+        verify(log, times(1)).warn("No image name configured for build stage: phase2. Image d2efc6645cb not tagged!");
+
+        verify(log, times(1)).info("Tagging container image 7f55eab001a from stage phase as registry.example.com/sample:0.2.1");
+        verify(log, times(0)).info("Tagging container image d2efc6645cbb6ea012f8adcaaab6b03ef847dd3d2b4fa418ca4cde57eff28a7f as registry.example.com/sample:0.2.1");
+
+        verify(log, times(1)).info("Built container image.");
+    }
+
+    @Test
     public void testMultiStageContainerFileWithMultilineOutputStep() throws MojoExecutionException, IOException, URISyntaxException {
         URI sampleBuildOutputUri = PushMojoTest.class.getResource("/multistagecontainerfile/samplebuildoutput_multiline_step.txt").toURI();
         Path sampleBuildOutputPath = Paths.get(sampleBuildOutputUri);

--- a/src/test/java/nl/lexemmens/podman/BuildMojoTest.java
+++ b/src/test/java/nl/lexemmens/podman/BuildMojoTest.java
@@ -544,6 +544,8 @@ public class BuildMojoTest extends AbstractMojoTest {
         // At random verify some lines
         verify(log, times(1)).debug("Processing line: 'STEP 1: FROM nexus.example:15000/adoptopenjdk/openjdk11:11.0.3 AS base'");
         verify(log, times(1)).debug("Processing candidate: 'STEP 7: LABEL Build-User=sample-user2 Git-Repository-Url=null'");
+        verify(log, times(1)).debug("Found a stage named: phase");
+        verify(log, times(1)).debug("Found a stage named: phase2");
 
         // Verify stage detection
         verify(log, times(1)).debug("Processing stage in Containerfile: base");

--- a/src/test/resources/multistagecontainerfile/Containerfile
+++ b/src/test/resources/multistagecontainerfile/Containerfile
@@ -5,10 +5,15 @@ ENV RUN_CMD="exec java -jar executable.jar"
 WORKDIR /application
 ENTRYPOINT /application
 
+# NOTE: DO NOT uncomment or remove below 'FROM' statement
+# FROM unknown AS phase
 
 FROM base as phase
 COPY target/executable.jar ./
 ENTRYPOINT ${RUN_CMD}
+
+# NOTE: DO NOT uncomment or remove below 'FROM' statement
+# FROM unknown as phase2
 
 FROM base as phase2
 COPY target/other-executable.jar ./

